cpp/src/arrow/array.cc

```diff
@@ -310,25 +310,35 @@ bool DecimalArray::IsNegative(int64_t i) const {
   return sign_bitmap_data_ != nullptr ? BitUtil::GetBit(sign_bitmap_data_, i) : false;
 }
 
-template <typename T>
-ARROW_EXPORT Decimal<T> DecimalArray::Value(int64_t i) const {
-  Decimal<T> result;
-  FromBytes(GetValue(i), &result);
-  return result;
-}
-
-template ARROW_EXPORT Decimal32 DecimalArray::Value(int64_t i) const;
-template ARROW_EXPORT Decimal64 DecimalArray::Value(int64_t i) const;
-
-template <>
-ARROW_EXPORT Decimal128 DecimalArray::Value(int64_t i) const {
-  Decimal128 result;
-  FromBytes(GetValue(i), IsNegative(i), &result);
-  return result;
+std::string DecimalArray::Value(int64_t i) const {
+  const auto type_ = std::dynamic_pointer_cast<DecimalType>(type());
+  const int precision = type_->precision;
+  const int scale = type_->scale;
+  const int byte_width = byte_width_;
+  const uint8_t* bytes = GetValue(i);
+  switch (byte_width) {
+    case 4: {
+      Decimal32 value;
+      FromBytes(bytes, &value);
+      return ToString(value, precision, scale);
+    }
+    case 8: {
+      Decimal64 value;
+      FromBytes(bytes, &value);
+      return ToString(value, precision, scale);
+    }
+    case 16: {
+      Decimal128 value;
+      FromBytes(bytes, IsNegative(i), &value);
+      return ToString(value, precision, scale);
```

Comment:
Can you add a namespace around these decimal methods, so that here it's `decimal::FromBytes` (i.e. `arrow::decimal::FromBytes`)?